File: client-library/vblob_cdn_client.py
"""
Python Client Library for the Relay file API

Targets /api/files/* (see /docs on your deployment). The older /api/v1/*
surface this client used to call is deprecated - see
app/lib/auth/api-auth.ts's withDeprecatedApiAuth.

Usage:
    from vblob_cdn_client import VBlobCDN

    cdn = VBlobCDN('your-api-key', 'https://your-domain.com')

    with open('image.png', 'rb') as f:
        result = cdn.upload(f, 'image.png', 'image/png')
        print(f'URL: {result["url"]}')

    files = cdn.list_files(q='invoice')
    info = cdn.get_file_info(result['shortId'])
    cdn.delete_file(result['id'])

Files above a few MB should use the multipart endpoints
(/api/files/multipart/init, batch-urls, complete, abort) documented at
/docs - this lightweight client only covers the single-shot upload path.
"""

import json
from typing import Optional, Dict, Any, BinaryIO, List, Union

try:
    import requests
except ImportError:
    raise ImportError("Install the 'requests' library: pip install requests")


class VBlobCDNError(Exception):
    """Exception raised for API errors."""

    def __init__(self, message: str, status_code: Optional[int] = None, response: Optional[Dict] = None):
        super().__init__(message)
        self.status_code = status_code
        self.response = response


class VBlobCDN:
    """Client for interacting with the Relay file API."""

    def __init__(self, api_key: Optional[str], base_url: str, timeout: int = 30):
        """
        Args:
            api_key: Your API key (optional - omit for anonymous requests)
            base_url: Base URL of your Relay deployment (e.g. 'https://your-domain.com')
            timeout: Request timeout in seconds (default: 30)
        """
        self.api_key = api_key
        self.base_url = base_url.rstrip('/')
        self.timeout = timeout
        self.session = requests.Session()
        if api_key:
            self.session.headers.update({'Authorization': f'Bearer {api_key}'})

    def _request(self, method: str, path: str, **kwargs) -> Dict[str, Any]:
        url = f'{self.base_url}{path}'
        response = self.session.request(method, url, timeout=self.timeout, **kwargs)

        try:
            data = response.json()
        except ValueError:
            data = {'error': response.text}

        if not response.ok or data.get('success') is False:
            raise VBlobCDNError(data.get('error', 'Request failed'), response.status_code, data)

        return data

    def upload(self, file: BinaryIO, filename: str, content_type: Optional[str] = None, folder_id: Optional[str] = None) -> Dict[str, Any]:
        """
        Upload a file in one call.

        Anonymous (no API key) uploads are capped at 25GB and expire after
        15 days; key-owned uploads are capped by the key's
        rateLimit.uploadSizeLimit and total storage quota.

        Returns the file record: id, name, size, url, viewUrl, mimeType,
        createdAt, isAnonymous, expiresAt, shortId.
        """
        files = {'file': (filename, file, content_type or 'application/octet-stream')}
        data = {}
        if folder_id:
            data['folderId'] = folder_id

        result = self._request('POST', '/api/files/upload', files=files, data=data)
        return result['data']

    def remote_upload(self, url: str, folder_id: Optional[str] = None) -> Dict[str, Any]:
        """Have the server fetch a URL and store it. Same limits as upload()."""
        body: Dict[str, Any] = {'url': url}
        if folder_id:
            body['folderId'] = folder_id
        result = self._request('POST', '/api/files/remote-upload', json=body)
        return result['data']

    def list_files(
        self,
        page: Optional[int] = None,
        limit: Optional[int] = None,
        folder_id: Optional[str] = None,
        q: Optional[str] = None,
        mime_type: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        List files uploaded with this API key. Requires the "list" permission.

        q: case-insensitive substring match on filename.
        mime_type: prefix match on MIME type, e.g. "image/".
        """
        params = {}
        if page is not None:
            params['page'] = page
        if limit is not None:
            params['limit'] = limit
        if folder_id is not None:
            params['folderId'] = folder_id
        if q is not None:
            params['q'] = q
        if mime_type is not None:
            params['mimeType'] = mime_type

        return self._request('GET', '/api/files/list', params=params)

    def get_file_info(self, short_id_or_ids: Union[str, List[str]]) -> List[Dict[str, Any]]:
        """Look up one or more files by their short ID (public - no API key needed)."""
        codes = ','.join(short_id_or_ids) if isinstance(short_id_or_ids, list) else short_id_or_ids
        result = self._request('GET', '/api/files/info', params={'file_code': codes})
        return result['result']

    def get_download_url(self, file_id: str) -> Dict[str, Any]:
        """Get a fresh presigned download URL for a file by its UUID (public - no API key needed)."""
        result = self._request('GET', f'/api/files/download/{file_id}')
        return result['data']

    def download_file(self, file_id: str) -> bytes:
        """Download a file's content."""
        info = self.get_download_url(file_id)
        response = requests.get(info['url'], timeout=self.timeout)
        if not response.ok:
            raise VBlobCDNError('Failed to download file', response.status_code)
        return response.content

    def delete_file(self, file_id: str, deletion_token: Optional[str] = None) -> None:
        """
        Delete a file by its UUID. Owner keys authenticate via the
        Authorization header; anonymous uploads need their deletion token
        instead.
        """
        params = {'fileId': file_id}
        if deletion_token:
            params['token'] = deletion_token
        self._request('DELETE', '/api/files/delete', params=params)

    def batch_delete_files(self, file_ids: List[str]) -> Dict[str, Any]:
        """Delete up to 100 key-owned files in one call. Requires the "delete" permission."""
        result = self._request('POST', '/api/files/batch-delete', json={'fileIds': file_ids})
        return result['data']

    def close(self):
        self.session.close()

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()


if __name__ == '__main__':
    import os

    api_key = os.environ.get('CDN_API_KEY')
    base_url = os.environ.get('CDN_BASE_URL', 'https://your-domain.com')

    if not api_key:
        print('Error: CDN_API_KEY environment variable not set')
        exit(1)

    with VBlobCDN(api_key, base_url) as cdn:
        files = cdn.list_files(limit=10)
        print(f'Total files: {files["pagination"]["total"]}')
        for f in files['data']:
            print(f'  - {f["name"]} ({f["size"]} bytes)')

File: client-library/vblob-cdn-client.ts
/**
 * TypeScript/JavaScript Client Library for the Relay file API
 *
 * Targets /api/files/* (see /docs on your deployment). The older
 * /api/v1/* surface this client used to call is deprecated - see
 * app/lib/auth/api-auth.ts's withDeprecatedApiAuth.
 *
 * Usage:
 * ```typescript
 * import { VBlobCDN } from './vblob-cdn-client';
 *
 * const cdn = new VBlobCDN('your-api-key', 'https://your-domain.com');
 *
 * const result = await cdn.upload(file);
 * console.log('Download URL:', result.url);
 *
 * const { data: files } = await cdn.listFiles({ q: 'invoice' });
 * const info = await cdn.getFileInfo(result.shortId);
 * await cdn.deleteFile(result.id);
 * ```
 *
 * Files above a few MB should use the multipart endpoints
 * (/api/files/multipart/init, batch-urls, complete, abort) documented at
 * /docs - this lightweight client only covers the single-shot upload path.
 */

export interface VBlobCDNConfig {
  apiKey?: string;
  baseUrl: string;
  timeout?: number;
}

export interface UploadResult {
  id: string;
  name: string;
  size: number;
  url: string;
  viewUrl?: string;
  mimeType: string;
  createdAt: string;
  isAnonymous: boolean;
  expiresAt: string | null;
  shortId: string;
}

export interface FileRecord {
  id: string;
  name: string;
  size: number;
  mime_type: string;
  path: string;
  short_id: string;
  folder_id: string | null;
  owner_id: string | null;
  is_anonymous: boolean;
  expires_at: string | null;
  created_at: string;
  updated_at: string;
  download_count: number;
}

export interface ListFilesOptions {
  page?: number;
  limit?: number;
  folderId?: string | null;
  includeFolders?: boolean;
  /** Case-insensitive substring match on filename. */
  q?: string;
  /** Prefix match on MIME type, e.g. "image/". */
  mimeType?: string;
}

export interface ListFilesResult {
  data: FileRecord[];
  pagination: { page: number; limit: number; total: number; totalPages: number };
  folders?: Array<{ id: string; name: string; parent_id: null; owner_id: null; created_at: string; updated_at: string }>;
}

export interface FileInfoResult {
  status: number;
  filecode: string;
  name?: string;
  size?: string;
  uploaded?: string;
  download?: string;
  status_field?: string;
}

export interface DownloadUrlResult {
  url: string;
  fileName: string;
  size: number;
  mimeType: string;
  expiresIn: number;
  expiresAt: string | null;
  downloads: number;
  shortId: string;
}

export interface BatchDeleteResult {
  results: Array<{ fileId: string; success: boolean; error?: string }>;
  deletedCount: number;
}

export class VBlobCDNError extends Error {
  constructor(
    message: string,
    public statusCode?: number,
    public response?: any
  ) {
    super(message);
    this.name = 'VBlobCDNError';
  }
}

export class VBlobCDN {
  private apiKey?: string;
  private baseUrl: string;
  private timeout: number;

  constructor(apiKey: string | undefined, baseUrl: string, timeout: number = 30000) {
    this.apiKey = apiKey;
    this.baseUrl = baseUrl.replace(/\/$/, '');
    this.timeout = timeout;
  }

  private authHeaders(): Record<string, string> {
    return this.apiKey ? { Authorization: `Bearer ${this.apiKey}` } : {};
  }

  private async request<T>(path: string, options: RequestInit = {}): Promise<T> {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);

    try {
      const response = await fetch(`${this.baseUrl}${path}`, {
        ...options,
        headers: { ...this.authHeaders(), ...options.headers },
        signal: controller.signal,
      });

      const data = await response.json();

      if (!response.ok || data.success === false) {
        throw new VBlobCDNError(data.error || 'Request failed', response.status, data);
      }

      return data;
    } catch (error) {
      if (error instanceof VBlobCDNError) throw error;
      if (error instanceof Error) throw new VBlobCDNError(error.message);
      throw new VBlobCDNError('Unknown error occurred');
    } finally {
      clearTimeout(timeoutId);
    }
  }

  /**
   * Upload a file. Anonymous (no API key) uploads are capped at 25GB and
   * expire after 15 days; key-owned uploads are capped by the key's
   * rateLimit.uploadSizeLimit and total storage quota.
   */
  async upload(file: File | Blob, filename?: string, folderId?: string): Promise<UploadResult> {
    const form = new FormData();
    const name = filename || (file instanceof File ? file.name : 'file');
    form.append('file', file, name);
    if (folderId) form.append('folderId', folderId);

    const result = await this.request<{ success: true; data: UploadResult }>('/api/files/upload', {
      method: 'POST',
      body: form,
    });
    return result.data;
  }

  /** Have the server fetch a URL and store it. Same limits as upload(). */
  async remoteUpload(url: string, folderId?: string): Promise<UploadResult> {
    const result = await this.request<{ success: true; data: UploadResult }>('/api/files/remote-upload', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ url, folderId }),
    });
    return result.data;
  }

  /** List files uploaded with this API key. Requires the "list" permission. */
  async listFiles(options: ListFilesOptions = {}): Promise<ListFilesResult> {
    const params = new URLSearchParams();
    if (options.page) params.set('page', String(options.page));
    if (options.limit) params.set('limit', String(options.limit));
    if (options.folderId !== undefined) params.set('folderId', options.folderId === null ? 'null' : options.folderId);
    if (options.includeFolders !== undefined) params.set('includeFolders', String(options.includeFolders));
    if (options.q) params.set('q', options.q);
    if (options.mimeType) params.set('mimeType', options.mimeType);

    const query = params.toString() ? `?${params.toString()}` : '';
    return this.request<ListFilesResult>(`/api/files/list${query}`);
  }

  /**
   * Look up one or more files by their short ID (public - no API key
   * needed). Accepts a single ID or an array for a batch lookup.
   */
  async getFileInfo(shortIdOrIds: string | string[]): Promise<FileInfoResult[]> {
    const codes = Array.isArray(shortIdOrIds) ? shortIdOrIds.join(',') : shortIdOrIds;
    const result = await this.request<{ result: FileInfoResult[] }>(
      `/api/files/info?file_code=${encodeURIComponent(codes)}`
    );
    return result.result;
  }

  /** Get a fresh presigned download URL for a file by its UUID (public - no API key needed). */
  async getDownloadUrl(fileId: string): Promise<DownloadUrlResult> {
    const result = await this.request<{ success: true; data: DownloadUrlResult }>(
      `/api/files/download/${encodeURIComponent(fileId)}`
    );
    return result.data;
  }

  /** Download a file's content as a Blob. */
  async downloadFile(fileId: string): Promise<Blob> {
    const { url } = await this.getDownloadUrl(fileId);
    const response = await fetch(url);
    if (!response.ok) throw new VBlobCDNError('Failed to download file', response.status);
    return response.blob();
  }

  /**
   * Delete a file by its UUID. Owner keys authenticate via the
   * Authorization header; anonymous uploads need their deletion token
   * instead (returned nowhere in this client - it's only shown once at
   * upload time via the raw API response for anonymous uploads).
   */
  async deleteFile(fileId: string, deletionToken?: string): Promise<void> {
    const params = new URLSearchParams({ fileId });
    if (deletionToken) params.set('token', deletionToken);
    await this.request(`/api/files/delete?${params.toString()}`, { method: 'DELETE' });
  }

  /** Delete up to 100 key-owned files in one call. Requires the "delete" permission. */
  async batchDeleteFiles(fileIds: string[]): Promise<BatchDeleteResult> {
    const result = await this.request<{ success: true; data: BatchDeleteResult }>('/api/files/batch-delete', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ fileIds }),
    });
    return result.data;
  }
}

// Node.js specific export
if (typeof module !== 'undefined' && module.exports) {
  module.exports = { VBlobCDN, VBlobCDNError };
}
